### PR TITLE
Add PG tag to every read option

### PR DESCRIPTION
**Pull request by jsotobroad**
Adding the option to not write out PG tags for MarkDuplicates and MergeBamAlignment.  Default behavior is kept the same as it was before but you can get speed increases if you set them appropriately.

### CODE REVIEW

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 1]
```diff
@@ -314,6 +315,11 @@ public void setMaxRecordsInRam(final int maxRecordsInRam) {
         this.maxRecordsInRam = maxRecordsInRam;
     }
 
+    /** Allows the caller to decide whether or not to write the PG tag to reads when applicable */
```
Comment (pshapiro4broad) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Where would be an appropriate place to document the default value of this property? Maybe here?

Comment (pshapiro4broad) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: For a docstring, I think it's better to describe behavior of the property, e.g. `Set addPGTagToReads. If true, the PG will be added to reads when ... If false, the PG tag will not be added.`

File: src/main/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 2]
```diff
@@ -354,7 +357,7 @@ protected int doWork() {
                     rec.getReadName().equals(opticalDuplicateQueryName) ||
                     recordInFileIndex == nextOpticalDuplicateIndex;
 
-            rec.setAttribute(DUPLICATE_TYPE_TAG, null);
+            if (CLEAR_DT) rec.setAttribute(DUPLICATE_TYPE_TAG, null);
```
Comment (pshapiro4broad) [inline src/main/java/picard/sam/markduplicates/MarkDuplicates.java]: I realize a lot of code here isn't written this way, but I think we should migrate to the Java coding standards in the picard code base. That means even a one-line `if` like this is written on two lines and has `{` `}` around its body.

File: src/main/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 3]
```diff
@@ -392,7 +395,7 @@ protected int doWork() {
             if (this.REMOVE_DUPLICATES            && rec.getDuplicateReadFlag()) continue;
             if (this.REMOVE_SEQUENCING_DUPLICATES && isOpticalDuplicate)         continue;
 
-            if (PROGRAM_RECORD_ID != null)  rec.setAttribute(SAMTag.PG.name(), chainedPgIds.get(rec.getStringAttribute(SAMTag.PG.name())));
+            if (PROGRAM_RECORD_ID != null && ADD_PG_TAG_TO_READS) rec.setAttribute(SAMTag.PG.name(), chainedPgIds.get(rec.getStringAttribute(SAMTag.PG.name())));
```
Comment (pshapiro4broad) [inline src/main/java/picard/sam/markduplicates/MarkDuplicates.java]: This `if` should also be written on two lines with `{ }`

File: src/main/java/picard/sam/util/ReadOutputCommandLineProgram.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package picard.sam.util;
+
+import org.broadinstitute.barclay.argparser.Argument;
+import picard.cmdline.CommandLineProgram;
+
+/**
+ * Abstract class that holds parameters and methods common to classes that want to add PG tags to reads in SAM/BAM files
+ *
+ */
+public abstract class ReadOutputCommandLineProgram extends CommandLineProgram {
```
Comment (pshapiro4broad) [inline src/main/java/picard/sam/util/ReadOutputCommandLineProgram.java]: Is `ReadOutput` the best name for this class? Would a typical picard developer expect to see the `ADD_PG_TAG` argument in it? Or maybe there's no good name for this concept?

File: src/main/java/picard/sam/markduplicates/util/AbstractOpticalDuplicateFinderCommandLineProgram.java [hunk 5]
```diff
@@ -27,14 +27,15 @@
...
+import picard.sam.util.ReadOutputCommandLineProgram;
 
 /**
  * Abstract class that holds parameters and methods common to classes that optical duplicate detection.  We put them here so that
  * the explanation about how read names are parsed is in once place
  *
  * @author Tim Fennell
  */
-public abstract class AbstractOpticalDuplicateFinderCommandLineProgram extends CommandLineProgram {
+public abstract class AbstractOpticalDuplicateFinderCommandLineProgram extends ReadOutputCommandLineProgram {
```
Comment (pshapiro4broad) [inline src/main/java/picard/sam/markduplicates/util/AbstractOpticalDuplicateFinderCommandLineProgram.java]: This provides a new command line argument for all subclasses. Is that option applicable for all subclasses? Or are only some subclasses affected by this argument?

File: src/main/java/picard/sam/util/ReadOutputCommandLineProgram.java (hunk 4 above)
Comment (jsotobroad) [inline src/main/java/picard/sam/util/ReadOutputCommandLineProgram.java]: I wasn't sure what would be a good name either.  This is something I discussed with @jacarey as well

File: src/main/java/picard/sam/markduplicates/MarkDuplicates.java (hunk 2 above)
Comment (jsotobroad) [inline src/main/java/picard/sam/markduplicates/MarkDuplicates.java]: 👍 I'll try to clean up wherever I see this if its something we want to move towards, all for it!

File: src/main/java/picard/sam/markduplicates/util/AbstractOpticalDuplicateFinderCommandLineProgram.java (hunk 5 above)
Comment (jsotobroad) [inline src/main/java/picard/sam/markduplicates/util/AbstractOpticalDuplicateFinderCommandLineProgram.java]: hmm `EstimateLibraryComplexity` inherits this but doesn't use it.  I cant see an obvious way to get only things that would use to inherit it but I'll see what I can do

**Comment by RoelKluin** [comment]
Thank you for this improvement, however, there are several tools that could also benefit from this feature. For instance, MarkDuplicatesWithMateCigar could be improved similarly. Also, can not writing the PG tag to every read be the default please? Besides speed improvement, I observe 25% larger files after markduplicates, and this might be the culprit. Please reopen.

**Comment by yfarjoun** [comment]
hi @RoelKluin,

you might want to either comment on the original issue (#902) or open a new issue. commenting on a merged PR will not have much effect otherwise. 

Alternatively, you are most welcome to submit a PR of your own, if you'd like!

Cheers,

Yossi.

**Pull request merged**